doc/VERSION.md

```diff
@@ -2,47 +2,84 @@
-is defined in [src/include/kdb.h](/src/include/kdb.h.in).
-Both applications and plugins use this API.
-
-Additionally there is also a very small API
-to plugins. It consists of only 5 functions
-and is described in [src/plugins/doc/doc.c](/src/plugins/doc/doc.c).
+## Scope
+
+The version applies to following parts of Elektra:
+
+- the API for programs using Elektra. Its interface
+  is declared in [src/include/kdb.h](/src/include/kdb.h.in).
+  Both applications and plugins use this API.
+- the high-level API as declared in
+  [src/include/elektra.h](/src/include/elektra.h).
+- the API to plugins as declared in
+  [src/plugins/doc/doc.h](/src/plugins/doc/doc.h).
+- the API to [libease](/src/include/kdbease.h)
+- the API to [libmeta](/src/include/kdbmeta.h)
+- the API to [libmerge](/src/include/kdbmerge.h)
+- the API to [module loading](/src/include/kdbmodule.h)
+- the API to [libnotification](/src/include/kdbnotification.h)
+- the API to [libopts](/src/include/kdbopts.h)
+- plugins that have `infos/status` of `compatible`.
+- the CLI tool `kdb` with its command-line
+  arguments, KDB access and its return values.
+
+## Behavior
+
+The following order defines the behavior of Elektra:
+
+1. [ABI](/tests/abi) test cases
+2. Non-ABI test cases (including Shell Recorder)
+3. The [API documentation](https://doc.libelektra.org/api/latest/html/)
+4. The [man pages](/doc/help)
+5. Tutorials (excluding Shell Recorder)
+6. Examples
+7. Implementation
+
+Any inconsistency within this artifacts constitutes a bug.
 
 ## Compatibility
 
 This section describes under which circumstances API
-and ABI incompatibilities may occur. As developer from
-Elektra your mission is to avoid that.
-The tool icheck against the interfaces mentioned
-above may help you too.
-
-In `0.8.*` the API and ABI must be always forward-compatible,
+and ABI incompatibilities may occur. As Elektra developer
+your mission is to avoid any unwanted incompatibilities.
+The tool icheck which checks the interfaces mentioned
+in the Scope may help you.
+
+Elektra uses a stricter version of
+[Semantic Versioning 2.0.0](https://semver.org/),
+with the extensions explained here.
+
+In `1.0.*` the API and ABI must be always forward-compatible
+and backwards-compatible.
+That means that Elektra's libraries may be upgraded and downgraded
+without any effect on applications, only bug or docu fixes
+are allowed.
+
+In `1.*` the API and ABI must be always forward-compatible,
 but not backwards-compatible.
-That means that a program written and compiled against 0.8.0
-compiles and links against 0.8.1. But because it is
+That means that a program written and compiled against 1.0.0
+compiles and links against 1.1.0. But because it is
 not necessarily backwards-compatible a program written
-for 0.8.1 may not link or compile against elektra 0.8.0
-(but it may do when you use the compatible subset, maybe
-with #ifdefs).
+for 1.1.0 may not link or compile against Elektra 1.0.0
```

Comment:
The ABI can be incompatible, even if compiling and linking works (e.g. changes to key names). We should note that. It would be better to specify that the program must also work as originally intended (apart from bug-fixes within Elektra).

Reply:
This sounds like very strong guarantee.

Strictly speaking if nothing else is specified, the implementation is the specification (as suggested in the Behavior chapter of this PR). Thus you cannot patch anything without changing the "original intention".

I think we would only be able to deliver a weak guarantee here that "we will try not to break" things intentionally. The perfect guarantee would be something we can automatically test.

Reply:
I missed the the implementation is on the list that specifies behaviour. I think that is a big mistake.  The implementation should never be part of the what defines the intended behaviour. This even makes bug-fixes impossible, since the bug is part of the implementation.

> "we will try not to break" things intentionally

Is very weak... Guaranteeing that compiling and linking works is fine, since it can be tested. But we should at least mention that this is not sufficient for a version to be ABI compatible.

Reply:
I changed at both places now, it is best if you look again if it suits now.

I think it is good and needed to provide strong guarantees. Obviously we are human and might fail from time to time.

That the implementation is on the list was an error on my part. Even alternative implementations of Elektra should not look at the implementation but instead open an issue if we did not test and document something properly.